src/AddressDerivation.hs

```diff
@@ -0,0 +1,215 @@
+--
+-- In theory, we should only consider two derivation types: soft and hard.
+--
+-- However, historically, addresses in Cardano used to be generated across the
+-- both soft and hard domain. We therefore introduce a 'WholeDomain' derivation
+-- type that is the exact union of `Hardened` and `Soft`.
+data DerivationType = Hardened | Soft | WholeDomain
+
+-- | An interface for doing hard derivations from the root private key
+class HardDerivation (key :: Depth -> * -> *) where
+    type AddressIndexDerivationType key :: DerivationType
+
+    -- | Derives account private key from the given root private key, using
+    -- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+    -- package for more details).
+    --
+    -- NOTE: The caller is expected to provide the corresponding passphrase (and
+    -- to have checked that the passphrase is valid). Providing a wrong passphrase
+    -- will not make the function fail but will instead, yield an incorrect new
+    -- key that doesn't belong to the wallet.
+    deriveAccountPrivateKey
+        :: ScrubbedBytes
+        -> key 'RootK XPrv
+        -> Index 'Hardened 'AccountK
+        -> key 'AccountK XPrv
+
+    -- | Derives address private key from the given account private key, using
+    -- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+    -- package for more details).
+    --
+    -- It is preferred to use 'deriveAddressPublicKey' whenever possible to avoid
+    -- having to manipulate passphrases and private keys.
+    --
+    -- NOTE: The caller is expected to provide the corresponding passphrase (and
+    -- to have checked that the passphrase is valid). Providing a wrong passphrase
+    -- will not make the function fail but will instead, yield an incorrect new
+    -- key that doesn't belong to the wallet.
+    deriveAddressPrivateKey
+        :: ScrubbedBytes
+        -> key 'AccountK XPrv
+        -> AccountingStyle
+        -> Index (AddressIndexDerivationType key) 'AddressK
+        -> key 'AddressK XPrv
+
+-- | An interface for doing soft derivations from an account public key
+class HardDerivation key => SoftDerivation (key :: Depth -> * -> *) where
+    -- | Derives address public key from the given account public key, using
+    -- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+    -- package for more details).
+    --
+    -- This is the preferred way of deriving new sequential address public keys.
+    deriveAddressPublicKey
+        :: key 'AccountK XPub
+        -> AccountingStyle
+        -> Index 'Soft 'AddressK
+        -> key 'AddressK XPub
+
+
+-- | Abstract interface for constructing a /Master Key/.
+class GenMasterKey (key :: Depth -> * -> *) where
+
+    -- | Generate a root key from a corresponding seed.
+    generateKeyFromSeed :: SomeMnemonic -> ScrubbedBytes -> key 'RootK XPrv
```

Comment:
Note that Icarus' and Ledger' style allows for defining an extra second factor passphrase. We could make this a type family as:

```hs
class GenMasterKey (key :: Depth -> * -> *) where
    type GenMasterKeyFrom key :: *

    -- | Generate a root key from a corresponding seed.
    genMasterKey :: GenMasterKeyFrom key -> ScrubbedBytes -> key 'RootK XPrv
```

Reply:
indeed, as thinking about Byron